coderunners/services.py

```diff
@@ -102,12 +104,17 @@ def check_code(code: dict[str, str], language: str, memory_limit: int, time_limi
     status = Status.OK if failed_test is None else test_results[failed_test].status
     max_memory = max(t.memory for t in test_results)
     max_time = max(t.time for t in test_results)
-
+    
+    if test_groups:
```

Comment:
It would be better to have a factory for this. This logic should be encapsulated in the `scoring` itself. `Checker` or `Compiler` use factories for this. Would be better to follow this logic consistently across the whole project.